src/Hazelcast.Net/Serialization/Compact/CompactReaderWriterBase.cs

```diff
@@ -0,0 +1,100 @@
+using Hazelcast.Core;
+
+namespace Hazelcast.Serialization.Compact
+{
+    internal abstract class CompactReaderWriterBase
+    {
+        protected readonly Schema Schema;
+        protected readonly int StartPosition;
+        protected readonly int DataStartPosition;
+
+        protected CompactReaderWriterBase(Schema schema, int startPosition)
+        {
+            Schema = schema ?? throw new ArgumentNullException(nameof(schema));
+
+            StartPosition = startPosition;
+
+            DataStartPosition = schema.HasReferenceFields
+                ? StartPosition + BytesExtensions.SizeOfInt
+                : StartPosition;
+        }
+
+        /// <summary>
+        /// Tries to get a field with the specified name (case-insensitive) and kind.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        /// <param name="fieldName">The actual field name, when found.</param>
+        /// <returns>Whether a field with the specified name (case-insensitive) was found.</returns>
+        public bool ValidateFieldNameInvariant(string name, [NotNullWhen(true)] out string? fieldName)
+        {
+            if (Schema.TryGetField(name, out var field, caseSensitive: false))
+            {
+                fieldName = field.FieldName;
+                return true;
+            }
+
+            fieldName = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="FieldKind"/> of a field.
+        /// </summary>
+        /// <param name="name">The name of the field.</param>
+        /// <returns>The <see cref="FieldKind"/> of the field, which can be <see cref="FieldKind.NotAvailable"/> if the field does not exist.</returns>
+        public FieldKind GetFieldKind(string name) => Schema.TryGetField(name, out var field) ? field.Kind : FieldKind.NotAvailable;
+
+        /// <summary>
+        /// Gets the schema field with the specified name (case-sensitive).
+        /// </summary>
+        /// <exception cref="SerializationException">No such field.</exception>
+        protected virtual SchemaField GetValidField(string name)
+        {
+            if (Schema.TryGetField(name, out var field)) return field;
+
+            throw new SerializationException($"Invalid field name \"{name}\" for schema {Schema}.");
+        }
+
+        /// <summary>
+        /// Gets the schema field with the specified name (case-sensitive) and kind.
+        /// </summary>
+        /// <exception cref="SerializationException">No such field.</exception>
+        protected virtual SchemaField GetValidField(string name, FieldKind kind)
+        {
+            var field = GetValidField(name);
+            if (field.Kind == kind) return field;
+
+            throw new SerializationException($"Invalid kind \"{kind}\" for field \"{name}\" of schema {Schema}, which is of kind \"{field.Kind}\".");
+        }
+
+        protected (int, byte) GetBooleanFieldPosition(string name)
```

Comment:
What does it make boolean special?

Reply:
Each integer, decimal, string, etc. is serialized as an independent field. OTOH non-nullable booleans are packed into bytes bit-fields. In other words, say you have 4 boolean fields, they will be serialized as 1 single byte, with each boolean corresponding to a bit in the byte. If you have 12 booleans you would have 2 bytes (8+4), etc. So booleans are special.